usr.bin/tr/tr.1
.\"	$OpenBSD: tr.1,v 1.26 2025/02/11 20:43:46 jmc Exp $
.\"	$NetBSD: tr.1,v 1.5 1994/12/07 08:35:13 jtc Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)tr.1	8.1 (Berkeley) 6/6/93
.\"
.Dd $Mdocdate: February 11 2025 $
.Dt TR 1
.Os
.Sh NAME
.Nm tr
.Nd translate characters
.Sh SYNOPSIS
.Nm tr
.Op Fl Ccs
.Ar string1 string2
.Nm tr
.Op Fl Cc
.Fl d
.Ar string1
.Nm tr
.Op Fl Cc
.Fl s
.Ar string1
.Nm tr
.Op Fl Cc
.Fl ds
.Ar string1 string2
.Sh DESCRIPTION
The
.Nm
utility copies the standard input to the standard output with substitution
or deletion of selected characters.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl C
Complements the set of characters in
.Ar string1 ;
for instance,
.Dq -C\ ab
includes every character except for
.Sq a
and
.Sq b .
.It Fl c
The same as
.Fl C .
.It Fl d
The
.Fl d
option causes characters to be deleted from the input.
.It Fl s
After all deletion and translation is completed,
the
.Fl s
option squeezes multiple consecutive occurrences of the characters
listed in the last operand (either
.Ar string1
or
.Ar string2 )
into a single instance of the character.
.El
.Pp
In the first synopsis form, the characters in
.Ar string1
are translated into the characters in
.Ar string2
where the first character in
.Ar string1
is translated into the first character in
.Ar string2
and so on.
If
.Ar string1
is longer than
.Ar string2 ,
the last character found in
.Ar string2
is duplicated until
.Ar string1
is exhausted.
.Pp
In the second synopsis form, the characters in
.Ar string1
are deleted from the input.
.Pp
In the third synopsis form, the characters in
.Ar string1
are compressed as described for the
.Fl s
option.
.Pp
In the fourth synopsis form, the characters in
.Ar string1
are deleted from the input, and the characters in
.Ar string2
are compressed as described for the
.Fl s
option.
.Pp
The following conventions can be used in
.Ar string1
and
.Ar string2
to specify sets of characters:
.Bl -tag -width [:equiv:]
.It character
Any character not described by one of the following conventions
represents itself.
.It \eoctal
A backslash followed by 1, 2, or 3 octal digits represents a character
with that encoded value.
To follow an octal sequence with a digit as a character, left zero-pad
the octal sequence to the full 3 octal digits.
.It \echaracter
A backslash followed by certain special characters maps to special
values.
.Pp
.Bl -tag -width "nn" -offset indent -compact
.It \ea
<alert character>
.It \eb
<backspace>
.It \ef
<form-feed>
.It \en
<newline>
.It \er
<carriage return>
.It \et
<tab>
.It \ev
<vertical tab>
.El
.Pp
A backslash followed by any other character maps to that character.
.It c-c
Represents the range of characters between the range endpoints, inclusively.
.It [:class:]
Represents all characters belonging to the defined character class.
Class names are:
.Pp
.Bl -tag -width "xdigit" -offset indent -compact
.It alnum
<alphanumeric characters>
.It alpha
<alphabetic characters>
.It blank
<blank characters>
.It cntrl
<control characters>
.It digit
<numeric characters>
.It graph
<graphic characters>
.It lower
<lower-case alphabetic characters>
.It print
<printable characters>
.It punct
<punctuation characters>
.It space
<space characters>
.It upper
<upper-case characters>
.It xdigit
<hexadecimal characters>
.El
.Pp
.\" All classes may be used in
.\" .Ar string1 ,
.\" and in
.\" .Ar string2
.\" when both the
.\" .Fl d
.\" and
.\" .Fl s
.\" options are specified.
.\" Otherwise, only the classes ``upper'' and ``lower'' may be used in
.\" .Ar string2
.\" and then only when the corresponding class (``upper'' for ``lower''
.\" and vice-versa) is specified in the same relative position in
.\" .Ar string1 .
.\" .Pp
With the exception of the
.Dq upper
and
.Dq lower
classes, characters
in the classes are in unspecified order.
In the
.Dq upper
and
.Dq lower
classes, characters are entered in
ascending order.
.Pp
For specific information as to which ASCII characters are included
in these classes, see
.Xr isalnum 3 ,
.Xr isalpha 3 ,
and related manual pages.
.It [=equiv=]
Represents all characters or collating (sorting) elements belonging to
the same equivalence class as
.Ar equiv .
If
there is a secondary ordering within the equivalence class, the characters
are ordered in ascending sequence.
Otherwise, they are ordered after their encoded values.
An example of an equivalence class might be
.Dq c
and
.Dq ch
in Spanish;
English has no equivalence classes.
.It [#*n]
Represents
.Ar n
repeated occurrences of the character represented by
.Ar # .
This
expression is only valid when it occurs in
.Ar string2 .
If
.Ar n
is omitted or is zero, it is interpreted as large enough to extend the
.Ar string2
sequence to the length of
.Ar string1 .
If
.Ar n
has a leading zero, it is interpreted as an octal value; otherwise,
it's interpreted as a decimal value.
.El
.Sh EXIT STATUS
.Ex -std tr
.Sh EXAMPLES
The following examples are shown as given to the shell:
.Pp
Create a list of the words in file1, one per line, where a word is taken to
be a maximal string of letters.
.Pp
.Dl $ tr -cs Qo [:alpha:] Qc Qo \en Qc < file1
.Pp
Translate the contents of file1 to upper-case.
.Pp
.Dl $ tr Qo [:lower:] Qc Qo [:upper:] Qc < file1
.Pp
Strip out non-printable characters from file1.
.Pp
.Dl $ tr -cd Qo [:print:] Qc < file1
.Sh SEE ALSO
.Xr sed 1
.Sh STANDARDS
The
.Nm
utility is compliant with the
.St -p1003.1-2008
specification,
except that the
.Fl C
option behaves the same as the
.Fl c
option since
.Nm
is not locale-aware.
.Pp
System V has historically implemented character ranges using the syntax
.Dq [c-c]
instead of the
.Dq c-c
used by historic
.Bx
implementations and
standardized by POSIX.
System V shell scripts should work under this implementation as long as
the range is intended to map in another range, i.e., the command
.Dq tr [a-z] [A-Z]
will work as it will map the
.Sq \&[
character in
.Ar string1
to the
.Sq \&[
character in
.Ar string2 .
However, if the shell script is deleting or squeezing characters as in
the command
.Dq tr\ -d\ [a-z] ,
the characters
.Sq \&[
and
.Sq \&]
will be
included in the deletion or compression list, which would not have happened
under an historic System V implementation.
Additionally, any scripts that depended on the sequence
.Dq a-z
to represent the three characters
.Sq a ,
.Sq - ,
and
.Sq z
will have to be rewritten as
.Dq a\e-z .
.Pp
The
.Nm
utility has historically not permitted the manipulation of NUL bytes in
its input and, additionally, has stripped NUL's from its input stream.
This implementation has removed this behavior as a bug.
.Pp
The
.Nm
utility has historically been extremely forgiving of syntax errors:
for example, the
.Fl c
and
.Fl s
options were ignored unless two strings were specified.
This implementation will not permit illegal syntax.
.Pp
It should be noted that the feature wherein the last character of
.Ar string2
is duplicated if
.Ar string2
has less characters than
.Ar string1
is permitted by POSIX but is not required.
Shell scripts attempting to be portable to other POSIX systems should use
the
.Dq [#*]
convention instead of relying on this behavior.


usr.bin/tr/tr.c
/*	$OpenBSD: tr.c,v 1.22 2022/12/04 23:50:49 cheloha Exp $	*/
/*	$NetBSD: tr.c,v 1.5 1995/08/31 22:13:48 jtc Exp $	*/

/*
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <err.h>

#include "extern.h"

int delete[NCHARS], squeeze[NCHARS];
int translate[NCHARS] = {
	0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,		/* ASCII */
	0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
	0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
	0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
	0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27,
	0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f,
	0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
	0x38, 0x39, 0x3a, 0x3b, 0x3c, 0x3d, 0x3e, 0x3f,
	0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47,
	0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f,
	0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57,
	0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
	0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67,
	0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f,
	0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77,
	0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f,
	0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87,
	0x88, 0x89, 0x8a, 0x8b, 0x8c, 0x8d, 0x8e, 0x8f,
	0x90, 0x91, 0x92, 0x93, 0x94, 0x95, 0x96, 0x97,
	0x98, 0x99, 0x9a, 0x9b, 0x9c, 0x9d, 0x9e, 0x9f,
	0xa0, 0xa1, 0xa2, 0xa3, 0xa4, 0xa5, 0xa6, 0xa7,
	0xa8, 0xa9, 0xaa, 0xab, 0xac, 0xad, 0xae, 0xaf,
	0xb0, 0xb1, 0xb2, 0xb3, 0xb4, 0xb5, 0xb6, 0xb7,
	0xb8, 0xb9, 0xba, 0xbb, 0xbc, 0xbd, 0xbe, 0xbf,
	0xc0, 0xc1, 0xc2, 0xc3, 0xc4, 0xc5, 0xc6, 0xc7,
	0xc8, 0xc9, 0xca, 0xcb, 0xcc, 0xcd, 0xce, 0xcf,
	0xd0, 0xd1, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7,
	0xd8, 0xd9, 0xda, 0xdb, 0xdc, 0xdd, 0xde, 0xdf,
	0xe0, 0xe1, 0xe2, 0xe3, 0xe4, 0xe5, 0xe6, 0xe7,
	0xe8, 0xe9, 0xea, 0xeb, 0xec, 0xed, 0xee, 0xef,
	0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7,
	0xf8, 0xf9, 0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff,
};

STR s1 = { STRING1, NORMAL, 0, OOBCH, { 0, OOBCH }, NULL, NULL };
STR s2 = { STRING2, NORMAL, 0, OOBCH, { 0, OOBCH }, NULL, NULL };

static void setup(int *, char *, STR *, int);
static void usage(void);

int
main(int argc, char *argv[])
{
	int ch, cnt, lastch, *p;
	int cflag, dflag, sflag;

	if (pledge("stdio", NULL) == -1)
		err(1, "pledge");

	cflag = dflag = sflag = 0;
	while ((ch = getopt(argc, argv, "Ccds")) != -1)
		switch(ch) {
		case 'C':
		case 'c':
			cflag = 1;
			break;
		case 'd':
			dflag = 1;
			break;
		case 's':
			sflag = 1;
			break;
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc < 1 || argc > 2)
		usage();

	/*
	 * tr -ds [-Cc] string1 string2
	 * Delete all characters (or complemented characters) in string1.
	 * Squeeze all characters in string2.
	 */
	if (dflag && sflag) {
		if (argc != 2)
			usage();

		setup(delete, argv[0], &s1, cflag);
		setup(squeeze, argv[1], &s2, 0);

		for (lastch = OOBCH; (ch = getchar()) != EOF;)
			if (!delete[ch] && (!squeeze[ch] || lastch != ch)) {
				lastch = ch;
				(void)putchar(ch);
			}
		exit(0);
	}

	/*
	 * tr -d [-Cc] string1
	 * Delete all characters (or complemented characters) in string1.
	 */
	if (dflag) {
		if (argc != 1)
			usage();

		setup(delete, argv[0], &s1, cflag);

		while ((ch = getchar()) != EOF)
			if (!delete[ch])
				(void)putchar(ch);
		exit(0);
	}

	/*
	 * tr -s [-Cc] string1
	 * Squeeze all characters (or complemented characters) in string1.
	 */
	if (sflag && argc == 1) {
		setup(squeeze, argv[0], &s1, cflag);

		for (lastch = OOBCH; (ch = getchar()) != EOF;)
			if (!squeeze[ch] || lastch != ch) {
				lastch = ch;
				(void)putchar(ch);
			}
		exit(0);
	}

	/*
	 * tr [-Ccs] string1 string2
	 * Replace all characters (or complemented characters) in string1 with
	 * the character in the same position in string2.  If the -s option is
	 * specified, squeeze all the characters in string2.
	 */
	if (argc != 2)
		usage();

	s1.str = (unsigned char *)argv[0];
	s2.str = (unsigned char *)argv[1];

	if (cflag)
		for (cnt = NCHARS, p = translate; cnt--;)
			*p++ = OOBCH;

	if (!next(&s2))
		errx(1, "empty string2");

	/* If string2 runs out of characters, use the last one specified. */
	ch = s2.lastch;
	if (sflag)
		while (next(&s1)) {
			translate[s1.lastch] = ch = s2.lastch;
			squeeze[ch] = 1;
			(void)next(&s2);
		}
	else
		while (next(&s1)) {
			translate[s1.lastch] = ch = s2.lastch;
			(void)next(&s2);
		}

	if (cflag)
		for (cnt = 0, p = translate; cnt < NCHARS; ++p, ++cnt)
			*p = *p == OOBCH ? ch : cnt;

	if (sflag)
		for (lastch = OOBCH; (ch = getchar()) != EOF;) {
			ch = translate[ch];
			if (!squeeze[ch] || lastch != ch) {
				lastch = ch;
				(void)putchar(ch);
			}
		}
	else
		while ((ch = getchar()) != EOF)
			(void)putchar(translate[ch]);
	exit (0);
}

static void
setup(int *table, char *arg, STR *str, int cflag)
{
	int cnt, *p;

	str->str = (unsigned char *)arg;
	bzero(table, NCHARS * sizeof(int));
	while (next(str))
		table[str->lastch] = 1;
	if (cflag)
		for (p = table, cnt = NCHARS; cnt--; ++p)
			*p = !*p;
}

static void
usage(void)
{
	fprintf(stderr,
	    "usage: tr [-Ccs] string1 string2\n"
	    "       tr [-Cc] -d string1\n"
	    "       tr [-Cc] -s string1\n"
	    "       tr [-Cc] -ds string1 string2\n");
	exit(1);
}
